File: python/trees/Solution94.py
# 94. 二叉树的中序遍历
from typing import Optional, List

from trees import TreeNode


class Solution:
    def inorderTraversalRecur(self, root: Optional[TreeNode]) -> List[int]:
        res = []

        def dfs(root, res):
            if root is None:
                return
            dfs(root.left, res)
            res.append(root.val)
            dfs(root.right, res)

        dfs(root, res)
        return res

    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        res = []
        stack = []
        while stack or root:
            if root:
                stack.append(root)
                root = root.left
            else:
                tmp = stack.pop()
                res.append(tmp.val)
                root = tmp.right
        return res

File: src/tree/Solution94.java
package tree;

import java.util.ArrayList;
import java.util.List;

// 94. 二叉树的中序遍历
public class Solution94 {

    public List<Integer> inorderTraversal(TreeNode root) {

        List<Integer> res = new ArrayList<>();
        inorder(root, res);
        return res;

    }

    private void inorder(TreeNode root, List<Integer> res) {
        if (root == null) return;
        inorder(root.left, res);
        res.add(root.val);
        inorder(root.right, res);
    }

}
